src/index.ts

```diff
@@ -1,69 +1,215 @@
+   * which determines the icon and label to display.
    */
   origin?: string;
+  /**
+   * The base URL of the app deployment form.
+   */
+  baseUrl?: string;
+  /**
+   * Additional query parameters to be appended to the baseUrl.
+   */
+  queryParameters?: Record<string, string>;
+}
+
+interface IPathWidget {
+  /**
+   * The context object associated with the widget.
+   * Contains the metadata and information about the widget.
+   */
+  context: {
+    /**
+     * The path of the current document or resource associated with the widget.
+     * Typically represents the file path or location of the notebook or document.
+     */
+    path: string;
+  };
+}
+
+const hasContextPath = (widget: any): widget is IPathWidget => {
+  return widget && widget.context && typeof widget.context.path === 'string';
+};
+
+export function buildQueryString(params: Record<string, string>): string {
+  return Object.entries(params)
+    .map(
+      ([key, value]) =>
+        `${encodeURIComponent(key)}=${encodeURIComponent(value)}`
+    )
+    .join('&');
 }
 
 const jhubAppsPlugin: JupyterFrontEndPlugin<void> = {
   id: 'jupyterlab-jhub-apps:commands',
   description: 'Adds additional commands used by jhub-apps.',
   autoStart: true,
-  activate: (app: JupyterFrontEnd) => {
-    const openURL = (url: string) => {
-      try {
-        window.open(url, '_blank', 'noopener,noreferrer');
-      } catch (error) {
-        console.warn(`Error opening ${url}: ${error}`);
-      }
-    };
-
+  optional: [ILayoutRestorer],
+  activate: (app: JupyterFrontEnd, restorer: ILayoutRestorer | null) => {
     const calculateIcon = (args: IDeployAppArgs) => {
       switch (args.origin) {
         case 'main-menu':
           return undefined;
         case 'context-menu':
+        case 'toolbar':
         case undefined:
         default:
           return deployAppIcon;
       }
     };
 
     app.commands.addCommand(CommandIDs.deployApp, {
-      execute: () => {
+      execute: ({
+        origin,
+        baseUrl = DEFAULT_BASE_URL,
+        queryParameters = {}
+      }: IDeployAppArgs) => {
         const currentWidget = app.shell.currentWidget;
-        const currentNotebookPath =
-          currentWidget && currentWidget instanceof DocumentWidget
-            ? currentWidget.context.path
-            : '';
-        let deployUrl;
-        if (currentNotebookPath !== '') {
-          deployUrl = `/services/japps/create-app?filepath=${encodeURIComponent(currentNotebookPath)}`;
+        const filepath = hasContextPath(currentWidget)
+          ? currentWidget.context.path
+          : '';
+
+        const selectedFilepath =
+          queryParameters.filepath !== undefined
+            ? queryParameters.filepath
+            : filepath;
+
+        const headless = (() => {
+          const value = queryParameters.headless;
+          if (value === undefined) {
+            return 'true';
+          }
+          if (
+            typeof value === 'string' &&
```

Comment:
Maybe a bit messy, but trying to deal with case where user passes an incorrect value (e.g. boolean).

Reply:
This looks fine, let's just move the logic to a standalone function (taking in a string, returning a boolean.